datafusion/physical-expr/src/equivalence.rs

```diff
@@ -885,6 +885,240 @@ fn req_satisfied(given: LexOrderingRef, req: &[PhysicalSortRequirement]) -> bool
+            );
+        }
+        _ => {}
+    };
+    Ok(right_oeq_class.normalize_with_equivalence_properties(join_eq_properties))
+}
+
+/// Calculate ordering equivalence properties for the given join operation.
+pub fn combine_join_ordering_equivalence_properties(
+    join_type: &JoinType,
+    left_oeq_properties: &OrderingEquivalenceProperties,
+    right_oeq_properties: &OrderingEquivalenceProperties,
+    schema: SchemaRef,
+    maintains_input_order: &[bool],
+    probe_side: Option<JoinSide>,
+    join_eq_properties: EquivalenceProperties,
+) -> Result<OrderingEquivalenceProperties> {
+    let mut new_properties = OrderingEquivalenceProperties::new(schema);
+    let left_columns_len = left_oeq_properties.schema().fields().len();
+    // All joins have 2 children
+    assert_eq!(maintains_input_order.len(), 2);
+    let left_maintains = maintains_input_order[0];
+    let right_maintains = maintains_input_order[1];
+    match (left_maintains, right_maintains) {
+        (true, true) => return plan_err!("Cannot maintain ordering of both sides"),
+        (true, false) => {
+            // In this special case, right side ordering can be prefixed with left side ordering.
+            if let (
+                Some(JoinSide::Left),
+                JoinType::Inner,
+                Some(left_oeq_class),
+                Some(right_oeq_class),
+            ) = (
+                probe_side,
+                join_type,
+                left_oeq_properties.oeq_class(),
+                right_oeq_properties.oeq_class(),
+            ) {
+                let updated_right_oeq = get_updated_right_ordering_equivalent_class(
+                    join_type,
+                    right_oeq_class,
+                    left_columns_len,
+                    &join_eq_properties,
+                )?;
+
+                // Right side ordering equivalence properties should be prepended with
+                // those of the left side while constructing output ordering equivalence
+                // properties since stream side is the left side.
+                //
+                // If the right table ordering equivalences contain `b ASC`, and the output
+                // ordering of the left table is `a ASC`, then the ordering equivalence `b ASC`
+                // for the right table should be converted to `a ASC, b ASC` before it is added
+                // to the ordering equivalences of the join.
+                let mut orderings = vec![];
+                for left_ordering in left_oeq_class.iter() {
+                    for right_ordering in updated_right_oeq.iter() {
+                        let mut ordering = left_ordering.to_vec();
+                        ordering.extend(right_ordering.to_vec());
+                        let ordering_normalized =
+                            join_eq_properties.normalize_sort_exprs(&ordering);
+                        orderings.push(ordering_normalized);
+                    }
+                }
```

Comment:
Previously we were prefixing with left ordering. We now prefix with all of the orderings inside left equivalence. We no longer use left output_ordering. This is the only change. in this function.